// @uiw/react-json-view/cjs/editor/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault")["default"];
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = require("react");
var _ = _interopRequireDefault(require("../"));
var _objectKey = require("./objectKey");
var _value = require("./value");
var _countInfoExtra = require("./countInfoExtra");
var _jsxRuntime = require("react/jsx-runtime");
var _excluded = ["onEdit", "components", "editable", "displayDataTypes", "onAdd", "onDelete"];
var JsonViewEditor = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var onEdit = props.onEdit,
    components = props.components,
    _props$editable = props.editable,
    editable = _props$editable === void 0 ? true : _props$editable,
    _props$displayDataTyp = props.displayDataTypes,
    displayDataTypes = _props$displayDataTyp === void 0 ? true : _props$displayDataTyp,
    onAdd = props.onAdd,
    onDelete = props.onDelete,
    reset = (0, _objectWithoutProperties2["default"])(props, _excluded);
  var comps = (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, components), {}, {
    countInfoExtra: function countInfoExtra(reprops) {
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(_countInfoExtra.CountInfoExtra, (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, reprops), {}, {
        editable: editable,
        onAdd: onAdd,
        onDelete: onDelete
      }));
    },
    objectKey: function objectKey(reprops) {
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(_objectKey.ObjectKey, (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, reprops), {}, {
        editableValue: editable,
        onEdit: onEdit,
        render: components === null || components === void 0 ? void 0 : components.objectKey
      }));
    },
    value: function value(reprops) {
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(_value.ReValue, (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, reprops), {}, {
        editableValue: editable,
        displayDataTypes: displayDataTypes,
        onDelete: onDelete,
        onEdit: onEdit
      }));
    }
  });
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_["default"], (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, reset), {}, {
    displayDataTypes: false,
    components: comps,
    ref: ref
  }));
});
JsonViewEditor.displayName = 'JVR.JsonViewEditor';
var _default = exports["default"] = JsonViewEditor;
module.exports = exports.default;
